includes/admin/class-edd-notices.php

```diff
@@ -19,242 +19,557 @@
-				switch( $_GET['edd-message'] ) {
-					case 'customer-deleted' :
-						$notices['updated']['edd-customer-deleted'] = __( 'Customer successfully deleted', 'easy-digital-downloads' );
-						break;
-					case 'user-verified' :
-						$notices['updated']['edd-user-verified'] = __( 'User successfully verified', 'easy-digital-downloads' );
-						break;
-					case 'email-added' :
-						$notices['updated']['edd-customer-email-added'] = __( 'Customer email added', 'easy-digital-downloads' );
-						break;
-					case 'email-removed' :
-						$notices['updated']['edd-customer-email-removed'] = __( 'Customer email removed', 'easy-digital-downloads');
-						break;
-					case 'email-remove-failed' :
-						$notices['error']['edd-customer-email-remove-failed'] = __( 'Failed to remove customer email', 'easy-digital-downloads');
-						break;
-					case 'primary-email-updated' :
-						$notices['updated']['edd-customer-primary-email-updated'] = __( 'Primary email updated for customer', 'easy-digital-downloads');
-						break;
-					case 'primary-email-failed' :
-						$notices['error']['edd-customer-primary-email-failed'] = __( 'Failed to set primary email', 'easy-digital-downloads');
-						break;
-				}
+		// Running Apache
+		if ( ! empty( $GLOBALS['is_apache'] ) && ! edd_htaccess_exists() ) {
+			$this->add_notice( array(
+				'id'             => 'edd-apache',
+				'class'          => 'error',
+				'is_dismissible' => false,
+				'message'        => array(
+					sprintf( __( 'The .htaccess file is missing from: %s', 'easy-digital-downloads' ), '<strong>' . $upload_directory . '</strong>' ),
+					sprintf( __( 'First, please resave the Misc settings tab a few times. If this warning continues to appear, create a file called ".htaccess" in the %s directory, and copy the following into it:', 'easy-digital-downloads' ), '<strong>' . $upload_directory . '</strong>' ),
+					sprintf( __( 'If you have already done this, you may safely and permenently %s.', 'easy-digital-downloads' ), '<a href="' . esc_url( add_query_arg( array( 'edd_action' => 'dismiss_notices', 'edd_notice' => 'uploads_is_public' ) ) ) . '">' . __( 'dismiss this notice', 'easy-digital-downloads' ) . '</a>' ),
+					'<pre>' . edd_get_htaccess_rules() . '</pre>'
+				)
+			) );
+		}
+	}
+
+	/**
+	 * Notices about data (migrations, etc...)
+	 *
+	 * @since 3.0
+	 */
+	private function add_data_notices() {
+
+		// Recount earnings
+		if ( class_exists( 'EDD_Recount_Earnings' ) ) {
+			$this->add_notice( array(
+				'id'             => 'edd-recount-earnings',
+				'class'          => 'error',
+				'is_dismissible' => false,
+				'message'        => sprintf( __( 'Easy Digital Downloads 2.5 contains a <a href="%s">built in recount tool</a>. Please <a href="%s">deactivate the Easy Digital Downloads - Recount Earnings plugin</a>', 'easy-digital-downloads' ), admin_url( 'edit.php?post_type=download&page=edd-tools&tab=general' ), admin_url( 'plugins.php' ) )
```

Comment:
While we are here, let's update this to read "Since Easy Digital Downloads 2.5, a built in recount tool..."

Reply:
Can you recommend the exact text? I can't make sense of how to combine what's here with your suggestion.